src/realm/query_expression.hpp

```diff
@@ -2012,45 +2019,75 @@ class Columns<Mixed> : public SimpleQuerySupport<Mixed> {
 public:
     using SimpleQuerySupport::evaluate; // don't hide the ObjKey overload
     using SimpleQuerySupport::SimpleQuerySupport;
-    void evaluate(size_t index, ValueBase& destination) override
+    void set_base_table(ConstTableRef table) override
+    {
+        SimpleQuerySupport::set_base_table(table);
+        m_ctrl.alloc = &get_link_map().get_target_table()->get_alloc();
+        m_ctrl.group = table->get_parent_group();
+    }
+
+    void reset_path(size_t index) final
     {
-        destination.init(false, 1); // Size of 1 is expected by SimpleQuerySupport
-        SimpleQuerySupport::evaluate(index, destination);
-        if (m_path.size() > 0) {
+        if (m_ctrl.path.size() > 0) {
+            Value<Mixed> destination;
+            destination.init(false, 1);
+            SimpleQuerySupport::evaluate(index, destination);
+
+            m_ctrl.matches.clear();
             if (auto sz = destination.size()) {
-                Collection::QueryCtrlBlock ctrl(m_path, *get_base_table(),
-                                                destination.m_from_list || !m_path_only_unary_keys);
                 for (size_t i = 0; i < sz; i++) {
-                    Collection::get_any(ctrl, destination.get(i), 0);
+                    Collection::get_any(m_ctrl, destination.get(i), 0);
                 }
-                sz = ctrl.matches.size();
-                destination.init(ctrl.from_list || sz == 0, sz);
-                destination.set(ctrl.matches.begin(), ctrl.matches.end());
+            }
+            if (m_ctrl.matches.empty()) {
+                // Make sure we at lease have one empty result
+                m_ctrl.matches.emplace_back();
             }
         }
+        m_destination_index = 0;
+    }
+
+    bool more() const final
+    {
+        return m_destination_index < m_ctrl.matches.size();
+    }
+
+    void evaluate(size_t index, ValueBase& destination) override
```

Comment:
It feels a bit weird to have two modes of evaluation like this where on one hand if the path is empty, the results are driven by the `index` parameter and the query node is stateless. Then in the other case, the `index` is ignored and the query node is advancing internal state at every call to evaluate. The other downside is that all query expressions have to pay the cost of two extra virtual function calls per evaluation with no benefit (`more()` and `reset_path()`).
What do you think about removing `more()` and `reset_path()` and then changing the design to something like this: 
```
struct QueryIndex {
    size_t index; // advanced directly for most queries
    size_t sub_index; // usually 0: used by many-to-many lists
    size_t sub_size; // usually 1: but can be set by the query node upon a call to evaluate
};
...

void evaluate(QueryIndex& index, ValueBase& destination) override {
    if (index.index == m_last_evaluated_index) {
        // fetch new lists
    }
    m_last_evaluated_index = index.index
    auto& matches = m_ctrl.matches[index.sub_index]; 
    // set destination as before
}
... 
// main driver loop also changes to iterate on index.index and while (index.sub_index < index.sub_size) 
```

Reply:
Thank you for this suggestion. I have a tendency to think that virtual functions is the answer to all problems.

Reply:
The updated design looks much cleaner to me, thank you!